compiler/src/iree/compiler/Codegen/Common/GenericVectorization.cpp

```diff
@@ -30,6 +32,80 @@ namespace mlir::iree_compiler {
 
 namespace {
 
+/// Folds IR resembling:
+/// ```
+///   %20 = vector.transfer_write %19, %16[%c0], %17 {in_bounds = [true]} 
+//      : vector<128xf16>, tensor<?xf16> 
+//    %21 = vector.transfer_read %20[%c0], %cst_2
+///     : tensor<?xf16>, vector<128xf16>
+/// ```
+/// into a simpler masked vector.transfer_read.
+/// After bufferization, this generally removes the need for materializing the
+/// write to memory.
+// TODO: Consider upstreaming
+struct FoldMaskedTransferRAW : OpRewritePattern<vector::TransferReadOp> {
+  using Base = OpRewritePattern<vector::TransferReadOp>;
+  using Base::Base;
+
+  LogicalResult matchAndRewrite(vector::TransferReadOp op,
+                                PatternRewriter &rewriter) const override {
+    // Fail to match if the read doesn't have pure tensor semantics.
+    if (!op.hasPureTensorSemantics())
+      return failure();
+
+    // Try to get the producing write op.
+    auto writeOp =
+        dyn_cast_or_null<vector::TransferWriteOp>(op.getBase().getDefiningOp());
+    // Fail to match if the write doesn't have pure tensor semantics.
+    if (!writeOp || !writeOp.hasPureTensorSemantics())
+      return failure();
+
+    Value valToStore = writeOp.getValueToStore();
+    // Fail to match if the in/out types are different
+    if (valToStore.getType() != op.getType())
+      return failure();
```

Comment:
Is this actually possible?

Reply:
Yes, one could read a bigger or smaller vector, there's no guarantees we are reading the same place.